cmd/kube-proxy/app/server_linux.go

```diff
@@ -381,48 +376,39 @@ func getConntrackMax(ctx context.Context, config proxyconfigapi.KubeProxyConntra
 	return 0, nil
 }
 
-func waitForPodCIDR(ctx context.Context, client clientset.Interface, nodeName string) (*v1.Node, error) {
-	// since allocators can assign the podCIDR after the node registers, we do a watch here to wait
-	// for podCIDR to be assigned, instead of assuming that the Get() on startup will have it.
-	ctx, cancelFunc := context.WithTimeout(ctx, timeoutForNodePodCIDR)
-	defer cancelFunc()
-
-	fieldSelector := fields.OneTermEqualSelector("metadata.name", nodeName).String()
-	lw := &cache.ListWatch{
-		ListFunc: func(options metav1.ListOptions) (object runtime.Object, e error) {
-			options.FieldSelector = fieldSelector
-			return client.CoreV1().Nodes().List(ctx, options)
-		},
-		WatchFunc: func(options metav1.ListOptions) (i watch.Interface, e error) {
-			options.FieldSelector = fieldSelector
-			return client.CoreV1().Nodes().Watch(ctx, options)
-		},
-	}
-	condition := func(event watch.Event) (bool, error) {
-		// don't process delete events
-		if event.Type != watch.Modified && event.Type != watch.Added {
+// getPodCIDRs returns PodCIDR for the node with the provided name. If required, it will wait
+// for the informer cache to have the node object for the node with the provided name, and
+// is populated with PodCIDRs.
+func getPodCIDRs(ctx context.Context, nodeLister corelisters.NodeLister, nodeName string) ([]string, error) {
+	return getPodCIDRsWithPollInterval(ctx, nodeLister, nodeName, time.Second)
+}
+
+// getPodCIDRsWithPollInterval implements getPodCIDRs with the given poll interval.
+func getPodCIDRsWithPollInterval(ctx context.Context, nodeLister corelisters.NodeLister, nodeName string, pollInterval time.Duration) ([]string, error) {
+	logger := klog.FromContext(ctx)
+	var podCIDRs []string
+	var node *v1.Node
+	var err error
+
+	pollErr := wait.PollUntilContextCancel(ctx, pollInterval, true, func(context.Context) (bool, error) {
```

Comment:
We don't need two very similar methods (`getPodIPs` and `getPodCIDRs`). We can just have one loop that you can tell it to wait for either just node IPs, or both node IPs and pod CIDRs. And the code could be in `NodeManager` rather than in `cmd/kube-proxy`.